core_dsl.go

```diff
@@ -501,6 +501,69 @@ func pushNode(node internal.Node, errors []error) bool {
 	return true
 }
 
+// NodeArgsTransformer is a hook which is called by the test construction DSL methods
+// before creating the new node. If it returns any error, the test suite
+// prints those errors and exits. The text and arguments can be modified,
+// which includes directly changing the args slice that is passed in.
+// Arguments have been flattened already, i.e. none of the entries in args is another []any.
+//
+// The node type is provided for information and remains the same.
+//
+// The offset is valid for calling NewLocation directly in the
+// implementation of TransformNodeArgs to find the location where
+// the Ginkgo DSL function is called. An additional offset supplied
+// by the caller via args is already included.
+//
+// A NodeArgsTransformer can be registered with AddTreeConstructionNodeArgsTransformer.
+type NodeArgsTransformer func(nodeType types.NodeType, offset Offset, text string, args []any) (string, []any, []error)
+
+// AddTreeConstructionNodeArgsTransformer registers a NodeArgsTransformer.
+// Only nodes which get created after registering a NodeArgsTransformer
+// are transformed by it. The returned function can be called to
+// unregister the transformer.
+//
+// Both may only be called during the construction phase.
+func AddTreeConstructionNodeArgsTransformer(transformer NodeArgsTransformer) func() {
```

Comment:
Instead of Push/Pop I went for Add/remove. That way the ordering of calls doesn't matter and one caller cannot accidentally pop the transformer pushed by another.

I intentionally left out locking because it shouldn't be needed.

Reply:
Initially I didn't want to guarantee any ordering (= not documented, iterating over map which forces randomness). That matches the random adding of transformers from different packages.

But for your use case with add/defer remove within a container node's body a deterministic order makes sense. I'm leaning towards "most recent first", as if it was on top of a stack.

Reply:
```
// If there is more than one registered transformer, then the most
// recently added ones get called first.
```

I suppose this can be resolved then.